source/opt/passes.h

```diff
@@ -74,6 +79,132 @@ class EliminateDeadConstantPass : public Pass {
+  // value of those spec constants can be determined, fold them to
+  // corresponding front-end constants.
+  bool ProcessImpl(ir::Module*);
+
+  // Try to get the result of the given OpSpecConstantOp CompositeExtract
+  // instruction. Returns a Value to be used to build a front-end constants
+  // if the result of can be determined. Otherwise return nullptr.
+  std::unique_ptr<Value> GetCompositeExtractResult(ir::Instruction* inst);
+
+  // Try to get the result of the given OpSpecConstantOp VectorShuffle
+  // instruction. Returns a Value to be used to build a front-end constants
+  // if the result of can be determined. Otherwise return nullptr.
+  std::unique_ptr<Value> GetVectorShuffleResult(ir::Instruction* inst);
+
+  // Try to get the result of the given OpSpecConstantOp instruction.  The spec
+  // opcode should indicate a scalar type operation or a broadcast operation
+  // for vector, e.g. IAdd. The spec code must obey SPIR-V spec. If the result
+  // constant has composite type, there will be additional scalar type
+  // constants need to be defined first and then used by the composite type
+  // result constants. A vector of the result ids of such scalar constants
+  // will be populated in this method and returned along with the result Value.
+  //
+  // If the result can be determined in the optimization, returns a tuple of
+  // result Value and a vector of ids that will be used to build the front-end
+  // constant of the result Value. Otherwise returns a tuple of nullptr and
+  // empty vector.
+  std::tuple<std::unique_ptr<Value>, std::vector<uint32_t>>
+  GetBroadcastOperationResult(ir::Instruction* inst);
+
+  // Inserts ONE instruction to define a front-end constant with the given
+  // result type result id at the location specified by the instruction
+  // iterator. This method lookup the id_to_const_val_ map to find the
+  // in-operands for the new instruction, determines the opcode based on the
+  // given type. This methods returns an instruction iterator which points to
+  // just inserted new constant instruction.
+  ir::Module::inst_iterator InsertConstantAt(
+      const analysis::Type* type, uint32_t result_id,
+      ir::Module::inst_iterator inst_iter);
+
+  // Returns a pointer to a Value object which holds the value extracted from
+  // the given instruction |inst| if |inst| defines a front-end constant with
+  // OpConstant or OpConstantComposite opcode, or it is a
+  // OpSpecConstantComposite instruction whose operands are all front-end
+  // constants. Otherwise returns a nullptr.
+  std::unique_ptr<Value> GetConstValue(ir::Instruction* inst);
+
+  // Returns a new ID to be added to the module.
+  uint32_t GetNewId() { return ++max_id_; }
+
+  // Returns the type of a given instruction. Returns nullptr if the
+  // instruction does not have a type id.
+  analysis::Type* GetType(const ir::Instruction* inst) {
+    if (inst->type_id() == 0) {
```

Comment:
I feel like we should make the interface of `type_mgr_->GetType()` more clear about passing in zero as parameter so that we can avoid the check here.